### ISSUE TITLE
How are the bins built? (related with missing representation)

### ISSUE THREAD

**Issue by BlindApe** [Microsoft/LightGBM#583]
Uniformly in range of variable? quantile based? 

There are some issues relate to this:

- How an outlier impact in overall bins distribution?
- How a manual recoding of missing by an extreme value (like -99999) impact in the rest of bins?
- Assigning missing by -99999 is a guarantee that there is a bin with all the missings and only them?

I think the double learning strategy of missing value in xgboost isn't so important. If the missing representation is a -99999 or +99999 greater enough, the process always can learn to split the first or last value. But in this case is important:
- The first or last bin has all the misssing and only them
- Introducing a big extreme number don't affect the rest of bins.

Package H2O gives the option of the method for bin construction (uniform, quantiles, ...). This article explain the effect in binning of introducing an extreme value (How does binning work for real/integer columns?):
[http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/gbm-faq/histograms_and_binning.html]

After read #122 I think reserve bin[0] for missing representation would be the best option. Separate missing is key in some datasets.

**Comment by guolinke** [comment]
@BlindApe 
Don't worry. The bin is built based on frequency. So a very big or small value doesn't have effect.
And actually, the missing values don't be used in bin building.

**Comment by Laurae2** [comment]
@BlindApe Outliers have no impact on binning (frequency based).

For the missing values double sided, they are important. See metric impact on 90%+ sparse dataset: https://github.com/Laurae2/gbt_benchmarks/issues/1#issuecomment-304289438

**Comment by BlindApe** [comment]
@Laurae2 Thank you, seems it had a big impact. 
But after read #516 I don't understand how use the new missing representation.
The key seems to manually recoding true zero to an small enough value, like 0.000001, but greater than
abs(value) > 1e-20f?
So all values that abs(value) < 1e-20f are considered missing and double tested for split on right and left. Lightgbm selects for which variables send NA to the min(feature) or max(feature).
Is right?

**Comment by Laurae2** [comment]
@BlindApe 

> I don't understand how use the new missing representation.

Missing value for `value` is when `abs(value) < 1e-20`.

> seems it had a big impact.

This one shows larger impact when using all data: https://github.com/Microsoft/LightGBM/pull/516#issuecomment-304543792

**Comment by guolinke** [comment]
@BlindApe you also can use NA to represent the missing value in R -python package.

**Comment by BlindApe** [comment]
@guolinke , but using NA, true zero are treated as missing too. If I've understood well, I will need to change 0 with 0.00001 if I want a different representation for zero and missings.

**Comment by Laurae2** [comment]
@BlindApe I don't think this will change much, 0s must be one side or another in any case (they would end up in the best scenario this way). In most cases, 0s have a special signification ("absence" of something), this is true in sparse formats like xgboost/LightGBM uses.
